apps/prairielearn/src/middlewares/clientFingerprint.ts

```diff
@@ -0,0 +1,42 @@
+import asyncHandler from 'express-async-handler';
+import * as sqldb from '@prairielearn/postgres';
+
+const sql = sqldb.loadSqlEquiv(__filename);
+
+export default asyncHandler(async (req, res, next) => {
+  const params = {
+    ip_address: req.ip,
+    user_id: res.locals.user.user_id,
+    session_id: req.session.id,
+    user_agent: req.headers['user-agent'],
+    accept_language: req.headers['accept-language'],
+    accept: req.headers['accept'],
+  };
+  // Need a better way to find or insert a fingerprint. Currently doing this through two calls but would probably be better to do it in one.
+  // First call, check if fingerprint exists in the database
+  let client_fingerprint_id: number | null =
+    parseInt(
+      (await sqldb.queryZeroOrOneRowAsync(sql.select_client_fingerprint, params)).rows[0]?.id,
+    ) ?? null;
```

Comment:
Rather than using `parseInt` here, I'd recommend using the Zod-powered query functions. In this case `queryOptionalRow` is probably what you want!